rfcs/20230704-dynamism-101.md

```diff
@@ -0,0 +1,826 @@
+there is one practical consideration worth discussing. Even though the proposed
+design doesn't involve `DynamicFooOp` operations, it may be worth it to
+introduce these operations (only in the StableHLO dialect) as an intermediate
+step. This would allow to roll out the new functionality (e.g.
+`DynamicReduceWindowOp`) sooner, without blocking on refactoring of the
+existing functionality (e.g. updating `ReduceWindowOp` as discussed above).
+
+<!-- markdownlint-disable line-length -->
+## <a id="p6" /> (P6) Represent shape computations as StableHLO operations on 0-dimensional tensors and drop support for unranked dynamism
+<!-- markdownlint-enable line-length -->
+
+**Baseline:** The StableHLO dialect represents sizes as 1-dimensional tensors,
+with static sizes expressed as `DenseIntElementsAttr` and dynamic sizes
+expressed as `1DTensorOf<[HLO_DimensionValue]>`.
+
+```c++
+class PadOp ... {
+  // This is how static sizes are represented in TableGen:
+  // I64ElementsAttr:$edge_padding_low
+  // I64ElementsAttr:$edge_padding_high
+  // I64ElementsAttr:$interior_padding
+  DenseIntElementsAttr getEdgePaddingLow();
+  DenseIntElementsAttr getEdgePaddingHigh();
+  DenseIntElementsAttr getInteriorPadding();
+}
+
+class DynamicPadOp ... {
+  // This is how dynamic sizes are represented in TableGen:
+  // HLO_DimensionTensor:$edge_padding_low
+  // HLO_DimensionTensor:$edge_padding_high
+  // HLO_DimensionTensor:$interior_padding
+  TypedValue<RankedTensorType> getEdgePaddingLow();
+  TypedValue<RankedTensorType> getEdgePaddingHigh();
+  TypedValue<RankedTensorType> getInteriorPadding();
+}
+```
+
+As mentioned in [F5](#f5), shape computations in StableHLO programs produce
+these 1-dimensional tensors using one of several approaches, including ops from
+Arith, Shape, StableHLO as well as other dialects. E.g. addition of dimension
+sizes can be represented via `stablehlo::AddOp`, `arith::AddIOp`, `shape::AddOp`
+and in a few other ways.
+
+Furthermore, as mentioned in [F4](#f4), using 1-dimensional tensors introduces
+considerable complexity because typical shape computations operate in terms of
+scalars. As a result, in code that constructs dynamic StableHLO ops, it is not
+unusual to encounter computations on 0-dimensional tensors whose results are
+then reshaped to 1-dimensional tensors and then concatenated together
+([example](https://github.com/tensorflow/mlir-hlo/blob/1fd13ef28d3f423363e59b1a80eb2c26e0c0979d/mhlo/transforms/chlo_legalize_to_hlo/chlo_legalize_to_hlo.cc#L1577-L1587)).
+
+**Proposal:** The RFC proposes to standardize on representing shape computations
+as StableHLO operations working 0-dimensional tensors. More specifically,
+the proposal is to:
+
+1) Affirm that only shape computations that use StableHLO operations are
+   supported in StableHLO portable artifacts (this can be changed in future
+   RFCs, but is out of scope for this one).
+2) Use `Variadic<0DTensorOf<HLO_DimensionValue>>` instead of
```

Comment:
Could you elaborate on using `HLO_DimensionValue` over e.g. `HLO_Int`?  Is the rationale for using `HLO_DimensionValue` that other dialects convert to `Index` instead of integer element type?

Following the proposal in [P5](#p5) ("variadic number of 0-dimensional tensors of integer type") verbatim, I'd think the proposal would use `Variadic<0DTensorOf<HLO_Int>>` instead. I'm asking since the StableHLO spec currently does not define `index` element type even though the StableHLO dialect uses it.

Reply:
>> Could you elaborate on using HLO_DimensionValue over e.g. HLO_Int? Is the rationale for using HLO_DimensionValue that other dialects convert to Index instead of integer element type?

That is true. For example, Shape [dialect](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Shape/IR/ShapeOps.td) replies heavily on index.

Reply:
"I'd think the proposal would use Variadic<0DTensorOf<HLO_Int>> instead". You're right. Using `HLO_DimensionValue` was an oversight on my part, this should be `HLO_Int`.

"Is the rationale for using HLO_DimensionValue that other dialects convert to Index instead of integer element type?". Yes, interoperability with other dialects was a factor in the design process, but in the end I decided to keep it out of scope of this RFC, and O2 goes into details about that. tl;dr is that switching to `index` would be a significant amount of work, which I think is best left for a follow-up RFC (this one is already huge).